### PR TITLE
Use global caching in JsonSerializerOptions

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Applies global caching for reflection metadata in two separate layers:

1. At the `MethodAccessor` layer caching generated dynamic methods.
2. Refactors the `JsonSerializerOptions` caching, using shared caching contexts shared across multiple `JsonSerializerOptions` instances "up to equivalence".

Contributes to #40072.

## Performance

Using benchmarks from https://github.com/dotnet/performance/pull/2236

|                   Method |        Job |                                                                                                        Toolchain |         Mean |        Error |       StdDev |       Median |          Min |          Max | Ratio | MannWhitney(3%) | RatioSD |  Gen 0 |  Gen 1 |  Gen 2 | Allocated | Alloc Ratio |
|------------------------- |----------- |----------------------------------------------------------------------------------------------------------------- |-------------:|-------------:|-------------:|-------------:|-------------:|-------------:|------:|---------------- |--------:|-------:|-------:|-------:|----------:|------------:|
|     CachedDefaultOptions | Job-IDHNJC | main |     742.3 ns |     55.64 ns |     64.08 ns |     730.1 ns |     618.9 ns |     837.6 ns |  1.00 |            Base |    0.00 | 0.0317 |      - |      - |     320 B |        1.00 |
|     CachedDefaultOptions | Job-VHZIGZ |   PR |     779.1 ns |     40.98 ns |     47.20 ns |     783.2 ns |     698.4 ns |     850.5 ns |  1.06 |            Same |    0.11 | 0.0292 |      - |      - |     320 B |        1.00 |
|                          |            |                                                                                                                  |              |              |              |              |              |              |       |                 |         |        |        |        |           |             |
|        NewDefaultOptions | Job-IDHNJC | main | 462,649.4 ns | 29,400.08 ns | 33,857.17 ns | 466,119.9 ns | 378,260.7 ns | 508,198.9 ns | 1.000 |            Base |    0.00 |      - |      - |      - |   13234 B |        1.00 |
|        NewDefaultOptions | Job-VHZIGZ |   PR |   1,281.0 ns |     76.38 ns |     87.96 ns |   1,285.5 ns |   1,109.7 ns |   1,449.0 ns | 0.003 |          Faster |    0.00 | 0.0586 | 0.0098 | 0.0049 |     637 B |        0.05 |
|                          |            |                                                                                                                  |              |              |              |              |              |              |       |                 |         |        |        |        |           |             |
|     NewCustomizedOptions | Job-IDHNJC | main | 450,203.9 ns | 26,126.89 ns | 29,039.97 ns | 453,785.3 ns | 408,028.4 ns | 502,898.1 ns | 1.000 |            Base |    0.00 |      - |      - |      - |   13317 B |        1.00 |
|     NewCustomizedOptions | Job-VHZIGZ |   PR |   1,402.5 ns |     85.25 ns |     98.18 ns |   1,383.5 ns |   1,245.4 ns |   1,577.6 ns | 0.003 |          Faster |    0.00 | 0.0631 | 0.0105 | 0.0053 |     662 B |        0.05 |
|                          |            |                                                                                                                  |              |              |              |              |              |              |       |                 |         |        |        |        |           |             |
|       NewCustomConverter | Job-IDHNJC | main | 388,271.0 ns | 18,375.55 ns | 19,661.63 ns | 387,544.6 ns | 348,863.4 ns | 426,233.6 ns |  1.00 |            Base |    0.00 |      - |      - |      - |   13356 B |        1.00 |
|       NewCustomConverter | Job-VHZIGZ |   PR |  17,386.3 ns |    970.16 ns |  1,078.34 ns |  17,761.1 ns |  15,454.0 ns |  19,390.9 ns |  0.04 |          Faster |    0.00 | 0.5618 | 0.0702 |      - |    6098 B |        0.46 |
|                          |            |                                                                                                                  |              |              |              |              |              |              |       |                 |         |        |        |        |           |             |
| NewCachedCustomConverter | Job-IDHNJC | main | 450,729.7 ns | 30,334.93 ns | 34,933.76 ns | 451,971.2 ns | 383,972.0 ns | 517,959.8 ns | 1.000 |            Base |    0.00 |      - |      - |      - |   13319 B |        1.00 |
| NewCachedCustomConverter | Job-VHZIGZ |   PR |   1,461.3 ns |     93.35 ns |    107.50 ns |   1,455.3 ns |   1,240.6 ns |   1,659.3 ns | 0.003 |          Faster |    0.00 | 0.0696 | 0.0107 | 0.0054 |     719 B |        0.05 |

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.Cache.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    internal sealed partial class ReflectionEmitCachingMemberAccessor
+    {
+        private sealed class Cache<TKey> where TKey : notnull
+        {
+            private int _invocationCount; // tracks total number of invocations to the cache; can be allowed to overflow.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.Cache.cs:16]: @stephentoub per offline discussion I did a couple of things here:

- Used an unbounded cache evicting entries based on sliding expiration. I ran a few benchmarks using a bounded MRU cache and which showed receding performance gains as the number of members grew relative to the fixed maximum cache size.
- Instead of using `Timer` to schedule evictions at a fixed interval, schedules evictions synchronously every _n_ th read attempt.

The obvious drawback of this approach is that eviction is iterating through all the dictionary context, which in certain access patterns can be expensive.

**Comment by eiriktsarpalis** [comment]
I've pushed a second commit that refactors the caching logic in `JsonSerializerOptions`. It hides all caches behind a `CachingContext` object that can be reused by multiple `JsonSerializerOptions` instances, provided that they are "equal" for a conservative notion of equality. Updated benchmarks:

|                     Method |        Job |                                                                                                        Toolchain |         Mean |       Error |      StdDev |       Median |          Min |          Max | Ratio | MannWhitney(3%) |  Gen 0 |  Gen 1 |  Gen 2 | Allocated | Alloc Ratio |
|--------------------------- |----------- |----------------------------------------------------------------------------------------------------------------- |-------------:|------------:|------------:|-------------:|-------------:|-------------:|------:|---------------- |-------:|-------:|-------:|----------:|------------:|
|          SerializeToString | Job-VVYEDL | main |     215.9 ns |     0.65 ns |     0.58 ns |     215.8 ns |     215.1 ns |     217.0 ns |  1.00 |            Base | 0.0246 |      - |      - |     256 B |        1.00 |
| SerializeToStringColdStart | Job-VVYEDL | main | 160,832.0 ns | 6,157.06 ns | 6,843.56 ns | 157,873.4 ns | 153,814.1 ns | 174,726.2 ns | 1.000 |            Base | 1.2376 | 0.6188 |      - |   12994 B |        1.00 |
| SerializeToStringColdStart | Job-GRJSXN |    PR |     713.6 ns |    13.52 ns |    12.65 ns |     711.8 ns |     699.4 ns |     742.8 ns | 0.004 |          Faster | 0.0532 | 0.0028 | 0.0028 |     597 B |        0.05 |

|                         Method |        Job |                                                                                                        Toolchain |         Mean |        Error |       StdDev |       Median |          Min |          Max | Ratio | MannWhitney(3%) | RatioSD |  Gen 0 |  Gen 1 |  Gen 2 | Allocated | Alloc Ratio |
|------------------------------- |----------- |----------------------------------------------------------------------------------------------------------------- |-------------:|-------------:|-------------:|-------------:|-------------:|-------------:|------:|---------------- |--------:|-------:|-------:|-------:|----------:|------------:|
|          DeserializeFromString | Job-VVYEDL | main |     326.0 ns |      3.36 ns |      2.63 ns |     326.1 ns |     321.3 ns |     329.6 ns |  1.00 |            Base |    0.00 | 0.0057 |      - |      - |      64 B |        1.00 |
| DeserializeFromStringColdStart | Job-VVYEDL | main | 247,420.2 ns | 18,488.01 ns | 21,290.82 ns | 237,937.0 ns | 227,630.9 ns | 305,346.6 ns | 1.000 |            Base |    0.00 | 1.0373 |      - |      - |   13136 B |        1.00 |
| DeserializeFromStringColdStart | Job-GRJSXN |  PR |     791.3 ns |     30.79 ns |     34.22 ns |     780.9 ns |     751.5 ns |     864.2 ns | 0.003 |          Faster |    0.00 | 0.0358 | 0.0030 | 0.0030 |     393 B |        0.03 |

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+namespace System.Text.Json.Serialization.Metadata
+{
+    internal sealed partial class ReflectionEmitCachingMemberAccessor : MemberAccessor
+    {
+        private static readonly ReflectionEmitMemberAccessor s_sourceAccessor = new();
+        private static readonly Cache<(string id, Type declaringType, MemberInfo? member)> s_cache =
+            new(slidingExpiration: TimeSpan.FromSeconds(5), evictionTriggerInterval: 200);
+
+        public override Action<TCollection, object?> CreateAddMethodDelegate<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] TCollection>()
+            => s_cache.GetOrAdd((nameof(CreateAddMethodDelegate), typeof(TCollection), null), static _ => s_sourceAccessor.CreateAddMethodDelegate<TCollection>());
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.cs:18]: @eerhardt this line is failing with the following trim analysis error:
```
Trim analysis error IL2091: System.Text.Json.Serialization.Metadata.ReflectionEmitCachingMemberAccessor.<>c__3<TCollection>.<CreateAddMethodDelegate>b__3_0(ValueTuple<String,Type,MemberInfo>): 'TCollection' generic argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods' in 'System.Text.Json.Serialization.Metadata.ReflectionEmitMemberAccessor.CreateAddMethodDelegate<TCollection>()'. The generic parameter 'TCollection' of 'System.Text.Json.Serialization.Metadata.ReflectionEmitCachingMemberAccessor.<>c__3<TCollection>' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```
I suspect it's because the `TCollection` parameter within the lambda parameter is lifted to a generic parameter in the compiler generated method, which doesn't have the necessary attribute annotation. I'm not sure how I could make this error go away without substantially refactoring the code. Any suggestions?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.cs]: Nevermind, TIL I can attach attributes to lambda literals.

Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.cs]: You may also be hitting the issue described here: https://github.com/dotnet/roslyn/issues/46646

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.Cache.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#if NETCOREAPP
+                    static (TKey key, Func<TKey, TValue> valueFactory) => new(valueFactory(key)),
+                    valueFactory);
+#else
+                    key => new(valueFactory(key)));
+#endif
+                long utcNowTicks = DateTime.UtcNow.Ticks;
+                Volatile.Write(ref entry.LastUsedTicks, utcNowTicks);
+
+                if (utcNowTicks - Volatile.Read(ref _lastEvictedTicks) > _evictionIntervalTicks)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.Cache.cs:41]:
```suggestion
                if (utcNowTicks - Volatile.Read(ref _lastEvictedTicks) >= _evictionIntervalTicks)
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.cs]: Looks like the one, thanks. I should link to that in the justification string.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    internal sealed partial class ReflectionEmitCachingMemberAccessor : MemberAccessor
+    {
+        private static readonly ReflectionEmitMemberAccessor s_sourceAccessor = new();
+        private static readonly Cache<(string id, Type declaringType, MemberInfo? member)> s_cache =
+            new(slidingExpiration: TimeSpan.FromSeconds(5), evictionInterval: TimeSpan.FromSeconds(1));
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitCachingMemberAccessor.cs:15]: > Could these values be hardcoded in the ctor since this is the only ctor reference?

What's the difference? I think it's better if these parameters are surfaced in the main implementation file.

> Also, how did you land on these values?

Good question. The purpose of the cache is to promote dynamic methods to long-living allocations that we want to eventually evict if unused (and since the cache is unbounded we don't need to evict on insertion). At the same time the eviction routine can be relatively expensive in the pathological scenaria we are looking to optimize (traversing through thousands of constantly reused member accessors) so it makes sense to spread it out as much as possible. This suggests to me that that both expiration and eviction intervals should be in the order of hundreds of milliseconds to a few seconds. The actual constants were chosen arbitrarily, although as might be expected choices doesn't demonstrate substantial variation in the context of microbenchmarks. We can always adjust these values in the future.

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/JsonSerializerContextTests.cs [hunk 5]
```diff
@@ -38,17 +38,13 @@ public static void Converters_AndTypeInfoCreator_NotRooted_WhenMetadataNotPresen
...
-                    // - Access JsonSerializerOptions._typeInfoCreationFunc
                     //
                     // If any of them changes, this test will need to be kept in sync.
 
                     // Confirm built-in converters not set.
                     AssertFieldNull("s_defaultSimpleConverters", optionsInstance: null);
                     AssertFieldNull("s_defaultFactoryConverters", optionsInstance: null);
 
-                    // Confirm type info dynamic creator not set.
-                    AssertFieldNull("_typeInfoCreationFunc", MetadataContext.Default.Options);
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/JsonSerializerContextTests.cs:50]: Ah thanks, I had that field removed initially and forgot to revert the change.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+                {
+                    return ctx;
+                }
+
+                lock (s_cache)
+                {
+                    if (s_cache.TryGetValue(options, out wr))
+                    {
+                        if (!wr.TryGetTarget(out ctx))
+                        {
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:142]: It's a technique called double-check locking, the first check avoids acquiring the lock in case of a cache hit.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+
+        internal void ClearCaches()
+        {
+            _cachingContext?.Clear();
+            _lastTypeInfo = null;
+        }
+
+        private void InitializeCachingContext()
+        {
+#if NETCOREAPP
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:76]: netstandard2.0 misses `System.HashCode`, but I guess I could polyfill with a naive implementation.

**Review by layomia — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+                    return ctx;
+                }
+            }
+
+            private static bool TryEvictDanglingEntries()
+            {
+                // Worst case scenario, the cache has been saturated with permanent entries.
+                // We want to avoid iterating needlessly through the entire cache every time
+                // a new options instance is created. For this reason we implement a backoff
+                // strategy to amortize the cost of eviction across multiple cache initializations.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:175]:
```suggestion
                // strategy to average out the cost of eviction across multiple cache initializations.
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+
+        /// <summary>
+        /// Defines a cache of CachingContexts; instead of using a ConditionalWeakTable which can be slow to traverse
+        /// this approach uses a regular dictionary pointing to weak references of <see cref="CachingContext"/>.
+        /// Relevant caching contexts are looked up using the equality comparison defined by <see cref="EqualityComparer"/>.
+        /// </summary>
+        private static class TrackedCachingContexts
+        {
+            private const int MaxTrackedContexts = 64;
+            private static readonly ConcurrentDictionary<JsonSerializerOptions, WeakReference<CachingContext>> s_cache =
```
Comment (vitek-karas) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:118]: I think this should make the cache work reasonably well with collectible assemblies, but it's hard to tell for sure. Could you please add a test which:
* Creates a new unloadable ALC and loads some assembly with test types into it
* Serializes/deserializes objects of the types from the assembly above
* Unloads the ALC
* Validates that the ALC was actually unloaded

Good guide what to do is here: https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability

In short what we want to avoid:
Global GC root which holds onto anything from unloadable ALCs as that will prevent the ALC from unloading. We already have quite a few caches in the FX which do this, so let's not add another one.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:118]: > In short what we want to avoid:
Global GC root which holds onto anything from unloadable ALCs as that will prevent the ALC from unloading. We already have quite a few caches in the FX which do this, so let's not add another one.

System.Text.Json already suffers from that issue since we're rooting the default JsonSerializerOptions instance (and any caches it may create):

https://github.com/dotnet/runtime/blob/cff924f9c823208b21d85f8fd5eed7eee8d4017a/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L35

We need to work out an eviction policy that takes into account the rooted instance as well, however that's not in scope for the current PR, which only addresses the performance issue.

Comment (vitek-karas) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:118]: Do we have a bug tracking this somewhere? (what I meant is that we should have a bug :wink:)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+            return _cachingContext.TryGetJsonTypeInfo(type, out typeInfo);
+        }
+
+        internal bool IsJsonTypeInfoCached(Type type) => _cachingContext?.IsJsonTypeInfoCached(type) == true;
+
+        /// <summary>
+        /// Return the TypeInfo for root API calls.
+        /// This has an LRU cache that is intended only for public API calls that specify the root type.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:54]: This was necessary?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+            public void Clear()
+            {
+                _converterCache.Clear();
+                _jsonTypeInfoCache.Clear();
+            }
+        }
+
+        /// <summary>
+        /// Defines a cache of CachingContexts; instead of using a ConditionalWeakTable which can be slow to traverse
+        /// this approach uses a regular dictionary pointing to weak references of <see cref="CachingContext"/>.
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:112]: "a regular dictionary"... it's a ConcurrentDictionary

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+                lock (s_cache)
+                {
+                    s_cache.Clear();
+                    s_recentEvictionCounts.Clear();
+                    s_evictionRunsToSkip = 0;
+                }
+            }
+
+            private static bool TryEvictDanglingEntries()
+            {
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:183]: Might be worth asserting the lock is held:
```C#
Debug.Assert(Monitor.IsEntered(s_cache));
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// Provides a conservative equality comparison for JsonSerializerOptions instances.
+        /// If two instances are equivalent, they should generate identical metadata caches;
+        /// the converse however does not necessarily hold.
+        /// </summary>
+        private class EqualityComparer : IEqualityComparer<JsonSerializerOptions>
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:255]: Nit: sealed

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+                hc.Add(options._allowTrailingCommas);
+                hc.Add(options._ignoreNullValues);
+                hc.Add(options._ignoreReadOnlyProperties);
+                hc.Add(options._ignoreReadonlyFields);
+                hc.Add(options._includeFields);
+                hc.Add(options._propertyNameCaseInsensitive);
+                hc.Add(options._writeIndented);
+                hc.Add(options._serializerContext);
+
+                foreach (JsonConverter converter in options.Converters)
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:324]: Iterating over an IList allocates an enumerator, which means this GetHashCode implementation is allocating, which you probably want to avoid.  It's also incurring multiple interface dispatch for MoveNext/Current.  Consider instead iterating via the IList indexer, iterating over the underlying `List<T>` if you can get access to it, or even leaving the converters out of the hash code entirely.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+    public sealed partial class JsonSerializerOptions
+    {
+        /// <summary>
+        /// Encapsulates all cached metadata referenced by the current <see cref="JsonSerializerOptions" /> instance.
+        /// Context can be shared across multiple equivalent options instances.
+        /// </summary>
+        private CachingContext? _cachingContext;
+
+        // Simple LRU cache for the public (de)serialize entry points that avoid some lookups in _cachingContext.
+        // Although this may be written by multiple threads, 'volatile' was not added since any local affinity is fine.
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:23]: Likewise, this has been moved over from a different source file, though I could submit a follow-up that corrects this.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+        /// </summary>
+        private sealed class CachingContext
+        {
+            private readonly ConcurrentDictionary<Type, JsonConverter> _converterCache = new();
+            private readonly ConcurrentDictionary<Type, JsonTypeInfo> _jsonTypeInfoCache = new();
+
+            public CachingContext(JsonSerializerOptions options)
+            {
+                Options = options;
+                _ = Count;
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:93]: `Count` is only accessed via reflection in testing. Added this to prevent the member from being removed in IL but maybe there's a better way to do this?

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:93]: `Count` is more expensive than you'd think, as it requires taking one or more locks.  We shouldn't be calling it in product src just to enable rooting the property for testing purposes.  This should probably have a debug-only trimmer xml file that keeps Count rooted only in debug, or if we really want to avoid that, a DynamicDependency from the ctor to Count (but that will still end up keeping the property rooted in production when it's only needed for testing).
cc: @eerhardt

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:93]: The way we do this is with a `ILLink.Descriptors.LibraryBuild.xml` file to preserve this property during our build. Here's our documentation:

https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/ILLink-files.md#illinkdescriptorsxml

See

https://github.com/dotnet/runtime/blob/a14d27ac7ff3a58e39d889db72f7ce682c7492fd/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.LibraryBuild.xml#L3-L5

as an example.

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:93]: @eerhardt what's the correct syntax for nested types? I tried adding a file via https://github.com/dotnet/runtime/pull/65376/commits/68bcb5384f01bba7e139380853cb6f1c7e44e389 but it looks like it's still trimming the property.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:93]: I guess it is a `/` 🤷 

See https://github.com/dotnet/linker/blob/main/docs/data-formats.md#preserve-a-type

**Comment by kunalspathak** [comment]
Improvements: https://github.com/dotnet/perf-autofiling-issues/issues/3641
